workflow/snakemake_rules/main_workflow.smk

```diff
@@ -1,7 +1,46 @@
+def _get_unaligned_sequence_file(wildcards):
+    if wildcards["origin"] == "":
+        if not isinstance(config["sequences"], str):
+            print("Error - a rule has asked for unaligned sequences without specifying an 'origin'. This necessitates config->sequences to be a string!")
+            # todo - the following exception is swallowed when snakemake decides that _running_ the rule which calls this fn (`_get_unaligned_sequence_file`)
+            # isn't necessary for the pipeline (e.g. due to intermediate files being present from a different run)
+            # how can we make snakemake exit here?
+            # Note that if the snakemake actually tries to run the rule, the exception causes an exit & message to be printed, as desired.
+            raise WorkflowError("Error - a rule has asked for unaligned sequences without specifying an 'origin'. This necessitates config->sequences to be a string!")
+        return config["sequences"]
+    origin = wildcards["origin"][1:] # trim leading `_`
+    if origin not in config["sequences"]:
+        print(f"Error - a rule has used an origin wildcard ({origin}) which doesn't have an associated sequences file!")
+        raise WorkflowError(f"Error - a rule has used an origin wildcard ({origin}) which doesn't have an associated sequences file!")
+    return config["sequences"][origin]
+
+def _get_filter_value(wildcards, key):
+    default = config["filter"].get(key, "")
+    if wildcards["origin"] == "":
+        return default
+    origin = wildcards["origin"][1:] # trim leading `_`
+    return config["filter"].get(origin, {}).get(key, default)
+
+
+def _get_aligned_sequence_file(wildcards):
+    """
+    Decide whether to use the aligned sequences file after diagnostics/refiltering,
+    or the aligned file without diagnostics
+    """
+    if wildcards["origin"] == "":
+        # todo - expose option to skip diagnostic / refiltering steps for normal (single-input) workflows?
+        return "results/aligned-filtered.fasta"
+    origin = wildcards["origin"][1:] # trim leading `_`
+    if config["filter"].get(origin, {}).get("skip_diagnostics", False):
+        return f"results/aligned{wildcards.origin}.fasta"
+    return f"results/aligned-filtered{wildcards.origin}.fasta"
+
+
+
 rule download_sequences:
     message: "Downloading sequences from S3 bucket {params.s3_bucket}"
     output:
-        sequences = config["sequences"]
+        sequences = config["default_sequences"]
```

Comment:
My biggest question / unknown at the moment is how to allow specifig inputs (e.g. based on the `origin` wildcard) to download filew from S3 -- both here and in the `download_preprocessed.smk`.

This specific solution is not great - if one of the input meadata / sequences files is `sequences.fasta` or `metadata.tsv` then it can be downloaded from AWS (because it matches the defaults). This allows our runs to continue as normal, but would break the situation where someone defined their own `config.sequences` or `config.metadata` filenames _and_ used the AWS download functionality.

In the case of ``download_preprocessed.smk`, if included, for single-input runs (`wildcards.origin=""`) these download rules will run (which is great) but I want to somehow allow _one_ of the multi-inputs to use these rules. In other words, make snakemake aware that one of the input FASTAs can be downloaded from S3 instead of being computed locally.

Reply:
I don't think I mind special casing `sequences.fasta` and `metadata.tsv`. As long as this is well documented I don't think it's a bad pattern.

Reply:
I think the solution in #550 achieves the desired goal in a much nicer way than here.

P.S. The downloaded sequences file isn't special cased to `sequences.fasta` but rather it will download to `config["sequences"]` if that file isn't available.